x/poolmanager/store.go

```diff
@@ -0,0 +1,598 @@
+
+	return registeredAlloyedPoolsIdArray, nil
+}
+
+// SetAllRegisteredAlloyedPoolIdArrayCached initializes the cache for the registered alloyed pools id array.
+func (k *Keeper) SetAllRegisteredAlloyedPoolIdArrayCached(ctx sdk.Context) error {
+	registeredAlloyPoolIds, err := k.GetAllRegisteredAlloyedPoolsIdArray(ctx)
+	if err != nil {
+		return err
+	}
+	k.cachedRegisteredAlloyedPoolIdArray = registeredAlloyPoolIds
+	return nil
+}
+
+//
+// Helpers
+//
+
+// queryAndCheckAlloyedDenom queries the smart contract for the alloyed denomination and validates its format.
+// It sends a query to the contract address to get the share denomination, then checks if the denomination
+// follows the expected format "factory/{contractAddr}/alloyed/{denom}".
+// Returns the alloyed denomination if valid, otherwise returns an error.
+func (k Keeper) queryAndCheckAlloyedDenom(ctx sdk.Context, contractAddr sdk.AccAddress) (string, error) {
+	queryBz := []byte(`{"get_share_denom": {}}`)
+	respBz, err := k.wasmKeeper.QuerySmart(ctx, contractAddr, queryBz)
+	if err != nil {
+		return "", err
+	}
+
+	var response alloyedpooltypes.ShareDenomResponse
+	err = json.Unmarshal(respBz, &response)
+	if err != nil {
+		return "", err
+	}
+	alloyedDenom := response.ShareDenom
+
+	parts := strings.Split(alloyedDenom, "/")
+	if len(parts) != 4 {
+		return "", types.InvalidAlloyedDenomFormatError{PartsLength: len(parts)}
+	}
+
+	if parts[0] != "factory" {
+		return "", types.InvalidAlloyedDenomPartError{PartIndex: 0, Expected: "factory", Actual: parts[0]}
+	}
+
+	if parts[1] != contractAddr.String() {
+		return "", types.InvalidAlloyedDenomPartError{PartIndex: 1, Expected: contractAddr.String(), Actual: parts[1]}
+	}
+
+	if parts[2] != "alloyed" {
+		return "", types.InvalidAlloyedDenomPartError{PartIndex: 2, Expected: "alloyed", Actual: parts[2]}
+	}
+
+	return alloyedDenom, nil
+}
+
+// snapshotTakerFeeShareAlloyComposition queries the smart contract for the total pool liquidity and calculates
+// the taker fee share agreements based on the liquidity of each asset in the pool. It returns a slice of
+// TakerFeeShareAgreement objects, each containing the denomination, skim percent, and skim address.
+// If the total alloyed liquidity is zero, it returns an error.
+func (k Keeper) snapshotTakerFeeShareAlloyComposition(ctx sdk.Context, contractAddr sdk.AccAddress) ([]types.TakerFeeShareAgreement, error) {
+	// Query for total pool liquidity
+	queryBz := []byte(`{"get_total_pool_liquidity": {}}`)
+	respBz, err := k.wasmKeeper.QuerySmart(ctx, contractAddr, queryBz)
+	if err != nil {
+		return []types.TakerFeeShareAgreement{}, err
+	}
+
+	var liquidityResponse alloyedpooltypes.TotalPoolLiquidityResponse
+	err = json.Unmarshal(respBz, &liquidityResponse)
+	if err != nil {
+		return []types.TakerFeeShareAgreement{}, err
+	}
+	totalPoolLiquidity := liquidityResponse.TotalPoolLiquidity
+
+	// Query for asset configs
+	queryBz = []byte(`{"list_asset_configs": {}}`)
```

Comment:
Is there a better way other than hard coding the request json here for liquidity and configs? Unclear to me if there is some other way I should be doing this.